ee/desktop/runner/runner.go

```diff
@@ -0,0 +1,424 @@
+// otherwise it returns the path to the current binary.
+func (r *DesktopUsersProcessesRunner) determineExecutablePath() (string, error) {
+	if r.executablePath != "" {
+		return r.executablePath, nil
+	}
+
+	executable, err := os.Executable()
+	if err != nil {
+		return "", fmt.Errorf("error getting executable path: %w", err)
+	}
+
+	return executable, nil
+}
+
+func (r *DesktopUsersProcessesRunner) userHasDesktopProcess(uid string) bool {
+	// have no record of process
+	proc, ok := r.uidProcs[uid]
+	if !ok {
+		return false
+	}
+
+	// have a record of process, but it died for some reason, log it
+	if !processExists(proc) {
+		level.Info(r.logger).Log(
+			"msg", "found existing desktop process dead for console user",
+			"pid", r.uidProcs[uid].process.Pid,
+			"process_path", r.uidProcs[uid].path,
+			"uid", uid,
+		)
+
+		return false
+	}
+
+	// have running process
+	return true
+}
+
+func processExists(processRecord processRecord) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+	defer cancel()
+
+	// the call to process.NewProcessWithContext ensures process exists
+	proc, err := process.NewProcessWithContext(ctx, int32(processRecord.process.Pid))
+	if err != nil {
+		return false
+	}
+
+	path, err := proc.ExeWithContext(ctx)
+	if err != nil || path != processRecord.path {
+		return false
+	}
+
+	return true
+}
+
+func (r *DesktopUsersProcessesRunner) processEnvVars(socketPath string) []string {
+	const varFmt = "%s=%s"
+	return append(
+		os.Environ(),
+		fmt.Sprintf(varFmt, "HOSTNAME", r.hostname),
+		fmt.Sprintf(varFmt, "AUTHTOKEN", r.authToken),
+		fmt.Sprintf(varFmt, "SOCKET_PATH", socketPath),
+	)
+}
+
+// socketPath returns standard pipe path for windows
+// on posix systems, it creates a folder and changes owner to the user
+// then provides a path to the socket in that folder
+func (r *DesktopUsersProcessesRunner) socketPath(uid string) (string, error) {
+	if runtime.GOOS == "windows" {
+		return fmt.Sprintf(`\\.\pipe\kolide_desktop_%s`, ulid.New()), nil
+	}
+
+	userFolderPath := filepath.Join(r.launcherRootDir, fmt.Sprintf("desktop_%s", uid))
+	if err := os.MkdirAll(userFolderPath, 0700); err != nil {
+		return "", fmt.Errorf("creating user folder: %w", err)
+	}
+
+	uidInt, err := strconv.Atoi(uid)
+	if err != nil {
+		return "", fmt.Errorf("converting uid to int: %w", err)
+	}
+
+	if err := os.Chown(userFolderPath, uidInt, -1); err != nil {
+		return "", fmt.Errorf("chowning user folder: %w", err)
+	}
+
+	path := filepath.Join(userFolderPath, "kolide_desktop.sock")
+	const maxSocketLength = 103
+	if len(path) > maxSocketLength {
+		return "", fmt.Errorf("socket path %s (length %d) is too long, max is %d", path, len(path), maxSocketLength)
+	}
+
+	return path, nil
+}
+
+func (r *DesktopUsersProcessesRunner) desktopCommand(executablePath, uid, socketPath string) (*exec.Cmd, error) {
+
+	cmd := exec.Command(executablePath, "desktop")
```

Comment:
General bias towards exec.CommandContext, with cancelable contexts. But maybe that's not as important here. Not sure?

Reply:
I think in this situation we do not want to use a context because we are managing the process ourselves. A call to the [cancel func ](https://pkg.go.dev/os/exec#CommandContext)sends an os.kill signal (which we do on interrupt) and we don't want a time out since the process is supposed to run indefinitely.

Reply:
👍 Makes sense to me